### ISSUE TITLE
xkbvalidate does not consult the services.xserver.xkbDir option

### ISSUE THREAD

**Issue by Josh-Tilles** [NixOS/nixpkgs#31138]
## Issue description
From what I can tell, [`xkbValidate`](https://github.com/NixOS/nixpkgs/blob/17.09/pkgs/tools/X11/xkbvalidate/xkbvalidate.c) does not use the directory pointed to by [`services.xserver.xkbDir`](https://nixos.org/nixos/options.html#services.xserver.xkbdir). As a result, `xkbValidate` believes my custom keyboard layout does not exist and fails the rebuild of NixOS.

(I’m not sure whether this is significant at all, but I noticed that NixOS uses libxkbcommon version 0.7.1 even though 0.7.2 is the most recent release.)

### Steps to reproduce
Clone [this repo](https://github.com/MerelyAPseudonym/custom-xkb-demo) as `/root/.local/share/X11/xkb/` and then ensure `/etc/nixos/configuration.nix` contains:
```nix
{
  # …

  services.xserver = {
    enable = true;
    layout = "us";
    xkbVariant = "dvrh";
    xbkDir = "/root/.local/share/X11/xkb/";
  };

  # …
}
```

Attempting to rebuild produces the following error:
```shell-session
[root@nixos:~]# nixos-rebuild build
building Nix...
building the system configuration...
these derivations will be built:
  /nix/store/9ywbdph9551s4k7ycikf9zrlijhpxrza-xkb-validated.drv
  /nix/store/dfqp1ryxfs1skgcb381k7fzcgk6ph14r-nixos-system-nixos-17.09.1927.6a0a00d91e.drv
building path(s) '/nix/store/6vhl471smmdlmkxrqzba290p52h35dmv-xkb-validated'
The XKB keyboard definition failed to compile:
    Couldn't process include statement for 'us(dvrh)'
    Abandoning symbols file "(unnamed)"
    Failed to compile xkb_symbols
    Failed to compile keymap
builder for '/nix/store/9ywbdph9551s4k7ycikf9zrlijhpxrza-xkb-validated.drv' failed with exit code 1
cannot build derivation '/nix/store/dfqp1ryxfs1skgcb381k7fzcgk6ph14r-nixos-system-nixos-17.09.1927.6a0a00d91e.drv': 1 dependencies couldn't be built
error: build of '/nix/store/dfqp1ryxfs1skgcb381k7fzcgk6ph14r-nixos-system-nixos-17.09.1927.6a0a00d91e.drv' failed
```

## Technical details
(generated by `nix-info --markdown`)
 - system: `"x86_64-linux"`
 - host os: `Linux 4.9.59, NixOS, 17.09.1927.6a0a00d91e (Hummingbird)`
 - multi-user?: `yes`
 - sandbox: `no`
 - version: `nix-env (Nix) 1.11.15`
 - channels(root): `"nixos-17.09.1927.6a0a00d91e"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

**Comment by andir** [comment]
I came across the same issue today while helping a friend migrating to NixOS..

Instead of pointing xkbDir to a path we created a derivation of `xkeyboard-config` which contained the additional files required. My current workaround will be to override the dependencies of `xkb-validate` so that the derived `xkeyboard-config` is used while the remaining parts of the system will just the standard version...
Not very satisfying but that is the easiest solution that I could come up with.

**Comment by nbp** [comment]
I am trying to add a new symbol file, but merging xkeyboard config with anything else is a nightmare, and replacing it with something else cause way too much recompilation to be practical.  This might be one of the cases where relying on a standard [ath makes sense, unless we could have each program compilation be independent of the xkb directory.

**Comment by rnhmjoj** [comment]
I'm working on a module/option that will (hopefully) allow to install a custom layout declaratively.
I found this [post](https://stesie.github.io/2016/08/nixos-custom-keyboard-layout) that describes the basic idea (how to avoid recompiling a ton of packages). I'm trying to inject a whole layout and not just a simple variant but it seems I can't get it right: the file is installed but setxkbmap doesn't load it and provides no helpful information at all.

This is the expression so far, if anyone want to try:
```nix
{ config, lib, pkgs, ... }:

let
  # options
  name = "<custom layout name>";
  description = "<custom layout long description>";
  languages = [ "eng" ];
  filepath = "<path to your layout>";

  xkeyboard_config_custom = lib.overrideDerivation pkgs.xorg.xkeyboardconfig (old: {
    buildInputs = old.buildInputs ++ [ pkgs.automake ];
    postPatch = ''
      # install layout file
      cp ${filepath} symbols/${name}

      # patch makefiles
      ${pkgs.ed}/bin/ed -v symbols/Makefile.am <<EOF
      /symbols_DATA =
      a
      ${name} \\
      .
      w
      EOF
      ${pkgs.ed}/bin/ed -v symbols/Makefile.in <<EOF
      /symbols_DATA =
      a
      ${name} \\
      .
      w
      EOF

      # add layout description
      ${pkgs.ed}/bin/ed -v rules/base.xml.in <<EOF
      /<\/layoutList>
      -
      a
      <layout>
        <configItem>
          <name>${name}</name>
          <_shortDescription>${name}</_shortDescription>
          <_description>${description}</_description>
          <languageList>
            ${lib.concatMapStrings (lang: "<iso639Id>${lang}</iso639Id>\n") languages}
          </languageList>
        </configItem>
        <variantList/>
      </layout>
      .
      w
      EOF
      cp rules/base.xml.in rules/base.evdev.xml.in
    '';
  });

in
{
  nixpkgs.config.packageOverrides = super: {
    xorg = super.xorg // rec {
      xorgserver = super.pkgs.lib.overrideDerivation super.xorg.xorgserver (old: {
        configureFlags = [
          "--enable-kdrive"
          "--enable-xephyr"
          "--enable-xcsecurity"
          "--with-default-font-path="
          "--with-xkb-bin-directory=${super.xorg.xkbcomp}/bin"
          "--with-xkb-path=${xkeyboard_config_custom}/share/X11/xkb"
          "--with-xkb-output=$out/share/X11/xkb/compiled"
          "--enable-glamor"
        ] ++ lib.optional super.stdenv.hostPlatform.isMusl "--disable-tls";

        postInstall = ''
          wrapProgram $out/bin/Xephyr \
            --set XKB_BINDIR "${xkbcomp}/bin" \
            --add-flags "-xkbdir ${xkeyboard_config_custom}/share/X11/xkb"
          wrapProgram $out/bin/Xvfb \
            --set XKB_BINDIR "${xkbcomp}/bin" \
            --set XORG_DRI_DRIVER_PATH ${super.mesa}/lib/dri \
            --add-flags "-xkbdir ${xkeyboard_config_custom}/share/X11/xkb"
          rm -fr $out/share/X11/xkb/compiled
          ( # assert() keeps runtime reference xorgserver-dev in xf86-video-intel and others
            cd "$dev"
            for f in include/xorg/*.h; do # */
              sed "1i#line 1 \"${old.name}/$f\"" -i "$f"
            done
          )
        '';
      });

      setxkbmap = super.pkgs.lib.overrideDerivation super.xorg.setxkbmap (old: {
        postInstall = ''
          mkdir -p $out/share
          ln -sfn ${xkeyboard_config_custom}/etc/X11 $out/share/X11
        '';
      });

      xkbcomp = super.pkgs.lib.overrideDerivation super.xorg.xkbcomp (old: {
        configureFlags = "--with-xkb-config-root=${xkeyboard_config_custom}/share/X11/xkb";
      });
    };
  };

  services.xserver.xkbDir = "${xkeyboard_config_custom}/etc/X11/xkb";
}

```

**Comment by ThomasMader** [comment]
There is now an updated version of the post. See http://stesie.github.io/2018/09/nixos-custom-keyboard-layout-revisited

**Comment by rnhmjoj** [comment]
@ThomasMader thank you! I finally get it to actually work. There is still a little problem though: if `i18n.consoleUseXkbConfig` is true the loadkeys command fails with a strange error: 

    too many (136) entries on one line

[custom-layout.nix](https://maxwell.ydns.eu/ipfs/QmXRNJb3Gxpjv1i7Y5SLHT8aKEtNLPfSmcjUQjU5afpUmx)

**Comment by rnhmjoj** [comment]
The error was specific to my layout, loadkeys didn't like some exotic unicode character.

**Comment by rnhmjoj** [comment]
The patch I mentioned in this thread has since made its way into NixOS as `services.xserver.extraLayouts` option. This does not fix the issue but provides a way to declaretively install custom layouts and xkbvalidate works in this case.

**Comment by attila-lendvai** [comment]
@rnhmjoj thanks a whole lot for this!

a question: is it normal that when i specify this, then a whole lot of packages get compiled locally on 20.09 stable? i mean stuff like the x server and gnome?

**Comment by nh2** [comment]
I think I wrote a commit that happens to fix this issue in PR #273239, would be happy if somebody could try it out.
